New Matrix Stats Aggregation module

nknize opened the pull request:
This PR adds a new aggs-matrix-stats module. The module introduces a new class of aggregations called `Matrix Aggregations`. Matrix aggregations work on multiple fields and produce a matrix as output. The matrix aggregation provided by this module is the `matrix_stats` aggregation which computes the following descriptive statistics over a set of fields:

* Covariance
* Correlation

For completeness (and interpretation purposes) the following per-field statistics are also provided:

* sample count
* population mean
* population variance
* population skewness
* population kurtosis

Example request:

```javascript
"aggs": {
    "matrixstats": {
        "matrix_stats": {
            "field": ["poverty", "income"]
        }
     }
}
```

Example response:
```javascript
{
	"aggregations": {
		"matrixstats": {
			"fields": [{
				"name": "income",
				"count": 50,
				"mean": 51985.1,
				"variance": 7.383377037755103E7,
				"skewness": 0.5595114003506483,
				"kurtosis": 2.5692365287787124,
				"covariance": {
					"income": 7.383377037755103E7,
					"poverty": -21093.65836734694
				},
				"correlation": {
					"income": 1.0,
					"poverty": -0.8352655256272504
				}
			}, {
				"name": "poverty",
				"count": 50,
				"mean": 12.732000000000001,
				"variance": 8.637730612244896,
				"skewness": 0.4516049811903419,
				"kurtosis": 2.8615929677997767,
				"covariance": {
					"income": -21093.65836734694,
					"poverty": 8.637730612244896
				},
				"correlation": {
					"income": -0.8352655256272504,
					"poverty": 1.0
				}
			}]
		}
	}
}
```

This PR replaces #16826. The differences include:

* renames aggregation from `multifield_stats` to `matrix_stats`
* adds aggregation code as a module instead of directly to core
* refactors Integration Tests as REST Tests

nknize commented:
This PR has been updated to include MultiValuesSource classes (per comment in #18285)

colings86 commented on core/src/main/java/org/elasticsearch/search/aggregations/InternalAggregation.java:
```diff
@@ -263,6 +265,7 @@ public final void readFrom(StreamInput in) throws IOException {
         public static final String FROM_AS_STRING = "from_as_string";
         public static final String TO = "to";
         public static final String TO_AS_STRING = "to_as_string";
+        public static final ParseField VALUE_TYPE = new ParseField("value_type");
```
I think this should actually be in ValuesSourceAggregatorBuilder since this `CommonFields` interface is for the response side of the aggregations but `value_type` is for the request side, so it doesn't really belong here

nknize replied:
Ah, so is `InternalAggregation` specifically for the response side? I'm not as intimate with the aggs framework so I thought it was just an internal framework class (request / response agnostic).

So then what do you think about having a `CommonFields` static class to `AggregatorBuilder` for common fields used on the request side?

colings86 replied:
+1 to having `CommonFields` class in `AggregatorBuilder`

colings86 commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStats.java:
```diff
@@ -0,0 +1,272 @@
+            return new InternalMatrixStats(name, 0, null, MatrixStatsResults.EMPTY(), pipelineAggregators(), getMetaData());
+        }
+
+        RunningStats runningStats = ((InternalMatrixStats) aggregations.get(0)).stats;
```
We had a bug in aggregations a while ago which was caused by using the first aggregation as the base for the reduce. Instead I think we should create a new matrixStats object here and then merge all of the shard results into it.

jpountz replied:
+1

colings86 commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsAggregator.java:
```diff
@@ -0,0 +1,150 @@
+                    if (valuesCount <= 0) {
+                        return null;
+                    }
+                    // get the field value (multi-value is the average of all the values)
```
is this a valid way to deal with multi-value fields (from a stats perspective)? If multi-value fields don't really work for this agg then maybe we should just stop it working on multi-value fields? (there are other aggs that do this)

jpountz replied:
+1

nknize replied:
Not really. In the end multi-value fields are treated as a single observation which is technically incorrect. The correct way is to factor multi-value fields into the weighted result. This is a todo (along with handling pair-wise missing values). But I'm not sure what you mean by "stop it working"? Do you mean skip the multi-value document? Or throw some kind of Unsupported exception?

jpountz replied:
I think thorowing an exception would be a good first step and then in a follow-up we could consider accepting a `mode` option like sorting does so that users can configure that we should take the min/max/median/average in case of multiple values

colings86 replied:
+1

colings86 commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsResults.java:
```diff
@@ -0,0 +1,242 @@
+    protected MatrixStatsResults(StreamInput in) {
+        try {
+            results = new RunningStats(in);
+            this.readFrom(in);
```
We don't really need this constructor and the readFrom() method. Can we get rid of the readFrom method and move the reading logic here? Also can we move the writeTo method so it's straight after this constructor so the read and write code is together? Also I think this constructor should throw IOException (I believe this is what other aggs do)

nknize replied:
Can't remove `readFrom()` since its part of the Interface. Do you want me to just noop?

nknize replied:
nvm. I'll convert from a `Streamable` to a `Writeable`. I just noticed the difference between the two.

colings86 commented:
@nknize I left some comments on the aggregation type stuff but I don't know enough about the actual statistics we are calculating here to review those bits. Maybe @brwe or @polyfractal would be able to review those bits?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/MatrixAggregationPlugin.java:
```diff
@@ -0,0 +1,52 @@
+
+public class MatrixAggregationPlugin extends Plugin {
+    static {
+        InternalMatrixStats.registerStreams();
```
should we rather do it in onModule? cc @colings86

colings86 replied:
+1

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStats.java:
```diff
@@ -0,0 +1,272 @@
+    public Double getMean(String field) {
+        if (results == null) {
+            return null;
+        }
```
to be consistent with the stats agg I think we should return NaN rather than null here (but still write null in toXContent since json does not support nan)

jpountz replied:
(and similarly for other below)

nknize replied:
+1 Good catch. The toXContent checks for null and returns NaN. Makes way more sense to return NaN here.

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStats.java:
```diff
@@ -0,0 +1,272 @@
+    @Override
+    public Map<String, HashMap<String, Double>> getCorrelation() {
+        return results.getCorrelations();
+    }
```
in terms of API I think we should choose which one of the two above methods we want to expose rather than exposing both

nknize replied:
+1

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStats.java:
```diff
@@ -0,0 +1,272 @@
+        final long count = in.readVLong();
+        if (count > 0) {
+            results = new MatrixStatsResults(in);
+        }
```
I am not happy that if stats or results is null on one side of the wire, then it becomes an empty instance on the other side of the wire. I would like it better if we either make sure that stats/results is never null or if we transport null so that null on one side is still null on the other side (I like the former better, but the latter would work for me too if keeping nulls is simpler)

nknize replied:
+1. I'm going to switch over to `Writeable` from `Streamable` and use the built-in `readOptionalWriteable` and `writeOptionalWriteable`. Looks like this should ensure consistency on both sides of the wire?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStats.java:
```diff
@@ -0,0 +1,272 @@
+    @Override
+    public InternalAggregation doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        // merge stats across all shards
+        aggregations.removeIf(p -> ((InternalMatrixStats)p).stats == null);
```
I think we should avoid modifying the list in place. Maybe make a copy first?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsAggregator.java:
```diff
@@ -0,0 +1,150 @@
+        boolean needsScores = false;
+        if (valuesSources != null) {
+            for (Map.Entry<String, ValuesSource.Numeric> valueSource : valuesSources.entrySet()) {
+                needsScores |= valueSource.getValue().needsScores();
```
if the key is not needed, let's just iterate over valuesSources.values()?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsAggregator.java:
```diff
@@ -0,0 +1,150 @@
+            private Map<String, Double> getFields(int doc) {
+                // get fieldNames to use as hash keys
+                ArrayList<String> fieldNames = new ArrayList<>(values.keySet());
+                HashMap<String, Double> fields = new HashMap<>(fieldNames.size());
```
this is called in `collect` which is typically called in a tight loop so we generally try to avoid having object creation done under collect(). Maybe we could use ordinal to identify field names so that the `HashMap<String, Double>` can become a simple `double[]` and then reuse the `double[]` across calls to collect?

polyfractal commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsAggregator.java:
```diff
@@ -0,0 +1,150 @@
+                            fieldValue += doubleValues.valueAt(i);
+                        }
+                    }
+                    fields.put(fieldName, fieldValue / valuesCount);
```
Unsure if this matters, but if `valueCount` is a positive, non-zero integer but all the values are `NaN`, the calculation will be 0.  Is that the intended behavior?  Since we're treating `NaN` as if it's a missing value and skip them, perhaps we should count how many values were collected instead of using `valuesCount`?

polyfractal commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/RunningStats.java:
```diff
@@ -0,0 +1,358 @@
+            nA2 = nA * nA;              // doc A num samples squared
+            nB2 = nB * nB;              // doc B num samples squared
+            // variance
+            variances.put(fieldName, varA + varB + d2 * nA * other.docCount / docCount);
```
This caught my eye at first because I wasn't sure how the order of operations was falling out, then I was curious.  Is the usage of delta squared mean here equivalent to a Pooled Variance like http://www.emathzone.com/tutorials/basic-statistics/combined-variance.html ?

I was expecting something like:

```
(
  nA * (Math.pow(varA,2) + Math.pow(meanA - combinedMean, 2))
  +
  nB * (Math.pow(varB,2) + Math.pow(meanB - combinedMean, 2))
) / docCount
```

But my math skills are *super rusty*, so heaps of salt etc etc :)

polyfractal replied:
Oh, just saw the PDF in the comment for streaming formulas.  I assume it's one of those, so you can probably ignore this :)

polyfractal commented:
I tried, and largely failed, to grok the stats calculations.  My math skills are pretty rusty and the streaming formulas look different from what I'm used to.  I'd try to get a second opinion from @brwe :)

nknize commented:
Great suggestions! I incorporated all feedback and the PR is ready for another review. Here are the main changes:

* Added `MultiValueMode` support for handling multi value fields (it was easier adding it here instead of postponing to a later PR)
* Added a `MultiValuesSource` class to encapsulate field names and ValuesSource arrays, this is now used to retrieve document field values instead of creating a Map inside of `collect`
* Updated `doReduce` to merge all shard results in an empty `RunningStats` object.
* Added Rest testing for `MultiValueMode` support

nknize commented:
Module documentation has been added and is ready for review /cc @clintongormley.  I think this feature is ready to go - pending a final review from @jpountz and/or @colings86.

nknize commented:
@rashidkpc or @spalger, the response format changed from the old #16826 PR. This was done to make it easier to parse (specifically for Kibana). Does this format look good to y'all?

spalger commented:
Thanks @nknize 

I appreciate that, but if I was going to be using this response I would probably transform it into something like this before using it:
```js
"aggregations": {
  "matrixstats": {
    "fields": [
      {
        "name": "income",
        "count": 50,
        "mean": 51985.1,
        "variance": 7.383377037755103E7,
        "skewness": 0.5595114003506483,
        "kurtosis": 2.5692365287787124,
        "covariance": {
          "poverty": -21093.65836734694
        },
        "correlation": {
          "poverty": -0.8352655256272504
        }
      },
      {
        "name": "poverty",
        "count": 50,
        "mean": 12.732000000000001,
        "variance": 8.637730612244896,
        "skewness": 0.4516049811903419,
        "kurtosis": 2.8615929677997767,
        "covariance": {
          "income": -21093.65836734694,
        },
        "correlation": {
          "income": -0.8352655256272504
        }
      }
    ]
  }
}
```

spalger commented:
Though honestly I'm not certain what all those values are... so guessing how I would use them it tricky. Either way, I don't see any point in optimizing this specific api response too much.

If response size is a concern then I think the current proposal makes sense, if not I would probably use a more hash-based style (rather than array's of values mapped to an array of keys by index)

spalger commented:
Something I said in #16826 may have lead you down this path?

> We'll have to transform it from object to array, but as long as the fields aren't in a meaningful order this is fine.

What I meant here was simply making an array of the fields so they could more easily be iterated over, as in the example response I posted.

nknize commented:
Thanks @spalger!

> Something I said in #16826 may have lead you down this path?

Ah. I think it was a combination of that and a separate discussion.

> The fact that the fields all had their values assigned as properties was preferred

@clintongormley, thoughts? Looks like its more useful to cut back to the first response format? It will get quite large with a lot of fields (this is only 2) but parsing is easier.

> guessing how I would use them it tricky

From a visualization perspective I envision a heatmap like in #16826 where you can visualize covariance or correlation. The rest of those values are just like `extended_stats`, likely displayed in a descriptive data table (or combined with a histogram).

georgebridgeman commented:
Hi all,

We visualise the correlation results using a [chord diagram](https://bost.ocks.org/mike/uberdata/), and also in a matrix (normally only showing the upper triangular). We also show ad-hoc correlation coefficients if the user or report is only interested in a few fields. Being able to key directly from one field to another to get the correlation coefficient like @spalger says saves having to figure array indexes.

I pulled @nknize 's branch a couple of weeks back (based on #16826) and the response format looked perfect. The array format is still workable but there's an extra step involved if picking out certain fields. 

Side note... I've been watching this PR for a while as it solves a major performance issue we've been having for a while. Currently, we effectively do what this aggregation does in code (using Apache Commons Math) after fetching all the documents. Having elasticsearch do this for us saves transporting the documents to the application server, making the visualisations much, much quicker. Thanks for all the work put into this PR!

nknize commented:
Thanks for the feedback @georgebridgeman! Wonderful to hear how this provides a performance boost for your use case. A chord diagram is another fantastic way to visualize the correlation/covariance results. We may want to consider adding it as a visualization option in kibana.

I'm going to go ahead and change the response format to look like the one posted by @spalger. It is quite a bit bigger but I think its more usable. If size becomes an issue we can consider alternatives in a future PR?

nknize commented:
Response format changed to the following:

```json
{
	"aggregations": {
		"matrixstats": {
			"fields": [{
				"name": "income",
				"count": 50,
				"mean": 51985.1,
                                "variance": 7.383377037755103E7,
                                "skewness": 0.5595114003506483,
                                "kurtosis": 2.5692365287787124,
                                "covariance": {
                                    "income": 7.383377037755103E7,
                                    "poverty": -21093.65836734694
                                },
                                "correlation": {
                                        "income": 1.0,
                                        "poverty": -0.8352655256272504
                                }
			}, {
				"name": "poverty",
				"count": 50,
                                "mean": 12.732000000000001,
                                "variance": 8.637730612244896,
                                "skewness": 0.4516049811903419,
                                "kurtosis": 2.8615929677997767,
                                "covariance": {
                                        "income": -21093.65836734694,
                                        "poverty": 8.637730612244896
                                },
                                "correlation": {
                                        "income": -0.8352655256272504,
                                        "poverty": 1.0
                                }
			}]
		}
	}
}
```

jpountz commented:
Even though I think it is more readable too, it might be more challenging to support scripts in the future since we would not have obvious keys. I don't think it is a blocker, this is experimental anyway so we'll be free to change the response format but I wanted to mention this potential annoyance with this format.

jpountz commented on docs/reference/modules/aggregations/matrix/stats.asciidoc:
```diff
@@ -0,0 +1,76 @@
+            "variance": [7.383377037755103E7, 8.637730612244896],
+            "skewness": [0.5595114003506483, 0.4516049811903419],
+            "kurtosis": [2.5692365287787124, 2.8615929677997767],
+            "covariance": [
```
doesn't look sync'ed with recent code changes?

jpountz commented on docs/reference/modules/aggregations/matrix/stats.asciidoc:
```diff
@@ -0,0 +1,76 @@
+    ...
+    "aggregations": {
+        "matrixstats": {
+            "field": ["income", "poverty"],
```
other APIs that take several fields such as the `query_string` query take `fields` rather than `field`, should we be consistent here?

nknize replied:
+1

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStats.java:
```diff
@@ -0,0 +1,46 @@
+    /** return the covariance between field x and field y */
+    Double getCovariance(String fieldX, String fieldY);
+    /** return the correlation coefficient of field x and field y */
+    Double getCorrelation(String fieldX, String fieldY);
```
s/Double/double/ since we now use NaN to represent empty results?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsAggregator.java:
```diff
@@ -0,0 +1,138 @@
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+                                                final LeafBucketCollector sub) throws IOException {
+        if (valuesSources == null || valuesSources.isEmpty()) {
```
no need to check for isEmpty since the constructor makes it null?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsAggregator.java:
```diff
@@ -0,0 +1,138 @@
+                    } else {
+                        stat.add(fieldNames, fieldVals);
+                    }
+                    stats.set(bucket, stat);
```
should we move this in the `if` block?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsResults.java:
```diff
@@ -0,0 +1,223 @@
+    /** object holding results - computes results in place */
+    final RunningStats results;
+    /** pearson product correlation coefficients */
+    protected HashMap<String, HashMap<String, Double>> correlation;
```
can you s/HashMap/Map/ and make it final?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsResults.java:
```diff
@@ -0,0 +1,223 @@
+    }
+
+    /** return the correlation coefficient between two fields */
+    public Double getCorrelation(String fieldX, String fieldY) {
```
I still think we should have this method or the above one but not both?

nknize replied:
I went ahead and made the other methods protected so they're still available to `InternalMatrixStats.getProperty`

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsResults.java:
```diff
@@ -0,0 +1,223 @@
+    }
+
+    /** return the kurtosis for the requested field */
+    public Double getKurtosis(String field) {
```
should it and other methods in this class return double rather than Double (and probably thorw an exception if a field is requested that was not specified at request time)?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/RunningStats.java:
```diff
@@ -0,0 +1,311 @@
+        init();
+    }
+
+    public RunningStats(final ArrayList<String> fieldNames, final double[] fieldVals) {
```
s/ArrayList/List/

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/RunningStats.java:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    /** updates running statistics with a documents field values **/
+    public void add(final ArrayList<String> fieldNames, final double[] fieldVals) {
```
it is weird that one comes as a list and the other one as an array, can we make both of them an array?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/RunningStats.java:
```diff
@@ -0,0 +1,311 @@
+    /** create an empty instance */
+    protected static RunningStats EMPTY() {
+        return new RunningStats();
+    }
```
when I see this method, it makes me think of a singleton that represents empty stats and should not be modified. Maybe we should remove this method and just expect callers to call the constructor?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/RunningStats.java:
```diff
@@ -0,0 +1,311 @@
+
+    /** updates running statistics with a documents field values **/
+    public void add(final ArrayList<String> fieldNames, final double[] fieldVals) {
+        if (fieldVals == null || fieldVals.length == 0) {
```
should it rather fail if fieldVals is null? and check that fieldNames and fieldVals have the same length?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSource.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    public boolean isEmpty() {
+        return values == null || values.length == 0;
```
it looks like it can never be null based on the constructor?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSourceAggregationBuilder.java:
```diff
@@ -0,0 +1,397 @@
+    private ValueType valueType = null;
+    private String format = null;
+    private Object missing = null;
+    private DateTimeZone timeZone = null;
```
do we make use of things like missing and timeZone already? If we don't I think we should remove them for now?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/RunningStats.java:
```diff
@@ -0,0 +1,311 @@
+    /** kurtosis values (fourth moment) */
+    protected HashMap<String, Double> kurtosis;
+    /** covariance values */
+    protected HashMap<String, HashMap<String, Double>> covariances;
```
to make this class more efficient, we might want to give an ordinal to every field and then use field ordinals as keys rather than field names? (I don't mind it being done as a follow-up, I'm just suggesting)

nknize replied:
+1 for followup PR. I'll open an issue

nknize commented:
Thanks for the latest round of review @jpountz. I went ahead and made the minor code changes, added support for missing values, and updated documentation. Let me know what you think.

jpountz commented on docs/reference/modules/aggregations/matrix/stats.asciidoc:
```diff
@@ -0,0 +1,112 @@
+}
+--------------------------------------------------
+
+The aggregation type is `matrix_stats` and the `field` setting defines the set of fields (as an array) for computing
```
s/field/fields/

jpountz commented on docs/reference/modules/aggregations/matrix/stats.asciidoc:
```diff
@@ -0,0 +1,112 @@
+`min`:: Pick the lowest value.
+`max`:: Pick the highest value.
+`sum`:: Use the sum of all values as sort value.
+`median`:: Use the median of all values as sort value.
```
maybe s/as sort values// since there is no sorting happening?

jpountz commented on docs/reference/modules/aggregations/matrix/stats.asciidoc:
```diff
@@ -0,0 +1,112 @@
+        "matrixstats": {
+            "matrix_stats": {
+                "fields": ["poverty", "income"],
+                "missing": [{"income" : 50000}] <1>
```
should it rather be `"missing": {"income" : 50000}`, ie. do we need the array?

jpountz commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsAggregator.java:
```diff
@@ -0,0 +1,137 @@
+                    final NumericDoubleValues doubleValues = values[i];
+                    final double value = doubleValues.get(doc);
+                    // skip if value is missing
+                    // todo: we could add option to specify a missing value?
```
the doc states there is such an option?

jpountz commented:
I'd like @colings86 to have a final look, at least for the core refactorings that you did. But otherwise this looks good to me for a first iteration!

colings86 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java:
```diff
@@ -76,6 +76,21 @@ public boolean valid() {
         return this;
     }
 
+    public ValuesSourceConfig<VS> format(final DocValueFormat format) {
```
Should we make all the fields of this class private now we have setters/getters for them so there is only one way of accessing/modifying the values? (maybe in a followup PR?)

nknize replied:
I went ahead and did it in this PR but will keep it as a separate commit when I merge.

colings86 commented:
@jpountz @nknize I took a look at the core changes and they LGTM. I also took a brief look at the module, there are unit and REST tests but we are missing integration tests for this. Is there a reason we have not done an integration tests for this aggregation? All other aggregations have integration tests.

nknize commented:
@colings86 I had asked @rjernst about integration testing in modules. I could be wrong but I think to avoid the overhead of starting/stopping integration test clusters in plugins and modules the `ESIntegTestCase` tests should be replaced by simple Unit and Rest tests? It was a bit tricky but what I did was refactor the integration tests (defined by `AbstractNumericTestCase`) from the original PR (#16826) to the yaml Rest definitions included in this PR. I also split out testing of `RunningStats` and `MatrixStatsResults` into a separate `RunningStatsTests` unit test class to ensure the RunningStats (and corresponding results) are computed as expected.

rjernst commented:
> there are unit and REST tests but we are missing integration tests for this

REST tests are real integration tests. No need for fantasy land tests.
